Jasmine/Tetris/ViewModels/TetrisGameViewModel.swift

```diff
@@ -1,4 +1,200 @@
+                    break
+                }
+                let newCoordinate = currentCoordinate.nextRow
+                tetrisGrid.addTile(at: newCoordinate, tileText: text)
+                coordinatesToShift.append((from: currentCoordinate, to: newCoordinate))
+            }
+        }
+        delegate?.animate(shiftTiles: coordinatesToShift)
+    }
+
+    // TODO: fetch from database to match valid phrase
+    private func getValidPhraseRange(_ line: String) -> CountableRange<Int>? {
+        let phrases = ["先发制人"]
+        for phrase in phrases {
+            let phraseLen = phrase.characters.count
+            for i in 0...line.characters.count - phraseLen {
+                let startIndex = line.index(line.startIndex, offsetBy: i)
+                if line[startIndex..<line.index(startIndex, offsetBy: phraseLen)] == phrase {
+                    return i..<i + phraseLen
+                }
+            }
+        }
+        return nil
+    }
+
+    // TODO: generate from database, base on existing grid
+    fileprivate func getRandomWord() -> String {
+        let words = "先发制人"
+        return String(words[words.index(words.startIndex,
+                                        offsetBy: Random.integer(toInclusive: words.characters.count))])
+    }
+}
+
+extension TetrisGameViewModel: TetrisGameViewModelProtocol {
+
+    var gameTitle: String {
+        return Constants.Game.Tetris.gameTitle
+    }
+
+    var gameInstruction: String {
+    	return Constants.Game.Tetris.gameInstruction
+    }
+
+    func canShiftFallingTile(to coordinate: Coordinate) -> Bool {
+        return !tetrisGrid.hasTile(at: coordinate)
+    }
+
+    func dropNextTile() -> (location: Coordinate, tileText: String) {
+        upcomingTiles.append(getRandomWord())
+        let tileText = upcomingTiles.removeFirst()
+        delegate?.redisplayUpcomingTiles()
+
+        fallingTileText = tileText
+        let randCol = Random.integer(toInclusive: Constants.Game.Tetris.columns)
+        return (location: Coordinate(row: Coordinate.origin.row, col: randCol),
+                tileText: tileText)
+    }
+
+    func landFallingTile(at coordinate: Coordinate) {
+        guard let fallingTileText = fallingTileText else {
+            assertionFailure("fallingTileText is not initialised")
+            return
+        }
+        tetrisGrid.addTile(at: coordinate, tileText: fallingTileText)
+
+        guard let destroyedCoordinates = checkForMatchingPhrase() else {
+            return
+        }
+        tetrisGrid.removeTiles(at: destroyedCoordinates)
+        delegate?.animate(destroyTilesAt: destroyedCoordinates)
+
+        currentScore += destroyedCoordinates.count
+
+        shiftDownTiles(destroyedCoordinates)
+    }
+
+    func swapFallingTile(withUpcomingAt index: Int) {
+        guard let currentFallingTileText = fallingTileText else {
+            assertionFailure("fallingTileText is not initialised")
```

Comment:
I use `assertionFailure` for methods with no/optional return value so that there's a possibility that the app can continue running in the event of a strange error.